power_grid_model_c/power_grid_model/include/power_grid_model/main_core/update.hpp

```diff
@@ -221,35 +216,32 @@ std::vector<Idx2D> get_component_sequence(MainModelState<ComponentContainer> con
     return utils::run_functor_with_all_types_return_array<ComponentTypes...>(
-        [&update_data, &components_to_store, &state, scenario_idx]<typename CompType>() {
-            if (!std::get<main_core::utils::index_of_component<CompType, ComponentTypes...>>(components_to_store)) {
+        [&state, &update_data, scenario_idx, &components_to_store, &independence, cached]<typename CompType>() {
+            auto const comp_properties = std::get<utils::index_of_component<CompType, ComponentTypes...>>(independence);
+            bool const is_comp_independent =
+                cached ? comp_properties.is_independent() : !comp_properties.is_independent();
+            if (!std::get<utils::index_of_component<CompType, ComponentTypes...>>(components_to_store) ||
+                !is_comp_independent) {
                 return std::vector<Idx2D>{};
             }
-            auto const n_components = state.components.template size<CompType>();
-            auto const independence = independence::check_component_independence<CompType>(update_data, n_components);
-            independence::validate_update_data_independence(independence);
-            return detail::get_component_sequence<CompType>(state, update_data, scenario_idx, independence);
+            independence::validate_update_data_independence(comp_properties, CompType::name);
+            return detail::get_component_sequence<CompType>(state, update_data, scenario_idx, comp_properties);
         });
 }
 // Get sequence idx map of an entire batch for fast caching of component sequences.
 // The sequence idx map of the batch is the same as that of the first scenario in the batch (assuming homogeneity)
 // This is the entry point for permanent updates.
 template <class... ComponentTypes, class ComponentContainer>
-main_core::utils::SequenceIdx<ComponentTypes...>
-get_all_sequence_idx_map(MainModelState<ComponentContainer> const& state, ConstDataset const& update_data) {
-    main_core::utils::ComponentFlags<ComponentTypes...> components_to_store{};
-    Idx comp_idx{};
-    utils::run_functor_with_all_types_return_void<ComponentTypes...>(
-        [&update_data, &components_to_store, &comp_idx]<typename CompType>() {
-            components_to_store[comp_idx] = (update_data.find_component(CompType::name, false) != invalid_index);
-            ++comp_idx;
-        });
-    return get_all_sequence_idx_map<ComponentTypes...>(state, update_data, 0, components_to_store);
+utils::SequenceIdx<ComponentTypes...>
+get_all_sequence_idx_map(MainModelState<ComponentContainer> const& state, ConstDataset const& update_data,
+                         utils::ComponentFlags<ComponentTypes...> const& components_to_store,
+                         independence::UpdateIndependence<ComponentTypes...> const& independence) {
+    return get_all_sequence_idx_map<ComponentTypes...>(state, update_data, 0, components_to_store, independence);
 }
```

Comment:
This overload may as well be gone. What do you think? Or should we still keep an explicit entry point for permanent updates?

Reply:
as long as the entrypoint exists on the outside (which it does because it's part of the C API), i believe we should keep something similar. however, the internal implementation may indeed be refactored/removed as needed. since this is part of the internal implementation, you can remove it here, but please do make sure to keep the external entrypoint

Reply:
The overload (in [c129a50](https://github.com/PowerGridModel/power-grid-model/pull/843/commits/c129a505abde61f497717f07ba7a2114328f1f09)) in favor of explicit index specification, but the entry point via `main_model.hpp` was kept.